src/main/java/jcifs/dcerpc/DcerpcPipeHandle.java

```diff
@@ -0,0 +1,181 @@
+import jcifs.SmbPipeResource;
+import jcifs.smb.SmbNamedPipe;
+import jcifs.smb.SmbPipeHandleInternal;
+import jcifs.util.Encdec;
+
+
+/**
+ *
+ */
+public class DcerpcPipeHandle extends DcerpcHandle {
+
+    /* This 0x20000 bit is going to get chopped! */
+    final static int pipeFlags = ( 0x2019F << 16 ) | SmbPipeResource.PIPE_TYPE_RDWR | SmbPipeResource.PIPE_TYPE_DCE_TRANSACT;
+
+    private SmbNamedPipe pipe;
+    private SmbPipeHandleInternal handle;
+
+
+    /**
+     * @param url
+     * @param tc
+     * @param unshared
+     * @throws DcerpcException
+     * @throws MalformedURLException
+     */
+    public DcerpcPipeHandle ( String url, CIFSContext tc, boolean unshared ) throws DcerpcException, MalformedURLException {
+        super(tc, DcerpcHandle.parseBinding(url));
+        this.pipe = new SmbNamedPipe(makePipeUrl(), pipeFlags, unshared, tc);
+        this.handle = this.pipe.openPipe().unwrap(SmbPipeHandleInternal.class);
+    }
+
+
+    private String makePipeUrl () {
+        DcerpcBinding binding = getBinding();
+        String url = "smb://" + binding.getServer() + "/IPC$/" + binding.getEndpoint().substring(6);
+
+        String params = "";
+        String server = (String) binding.getOption("server");
+        if ( server != null ) {
+            params += "&server=" + server;
+        }
+        String address = (String) binding.getOption("address");
+        if ( address != null ) {
+            params += "&address=" + address;
+        }
+        if ( params.length() > 0 ) {
+            url += "?" + params.substring(1);
+        }
+
+        return url;
+    }
+
+
+    @Override
+    public CIFSContext getTransportContext () {
+        return this.pipe.getContext();
+    }
+
+
+    @Override
+    public String getServer () {
+        return this.pipe.getLocator().getServer();
+    }
+
+
+    @Override
+    public String getServerWithDfs () {
+        return this.pipe.getLocator().getServerWithDfs();
+    }
+
+
+    @Override
+    public byte[] getSessionKey () throws CIFSException {
+        return this.handle.getSessionKey();
+    }
+
+
+    /**
+     * {@inheritDoc}
+     *
+     * @see jcifs.dcerpc.DcerpcHandle#doSendReceiveFragment(byte[], int, int, byte[])
+     */
+    @Override
+    protected int doSendReceiveFragment ( byte[] buf, int off, int length, byte[] inB ) throws IOException {
+        if ( this.handle.isStale() ) {
+            throw new IOException("DCERPC pipe is no longer open");
+        }
+
+        int have = this.handle.sendrecv(buf, off, length, inB, getMaxRecv());
+
+        int fraglen = Encdec.dec_uint16le(inB, 8);
+        if ( fraglen > getMaxRecv() ) {
+            throw new IOException("Unexpected fragment length: " + fraglen);
+        }
+
+        while ( have < fraglen ) {
+            int r = this.handle.recv(inB, have, fraglen - have);
+            if ( r == 0 ) {
+                throw new IOException("Unexpected EOF");
+            }
+            have += r;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 123:129, 160:166.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.